// @azure/cosmos/lib/src/queryExecutionContext/EndpointComponent/TopEndpointComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @hidden */
class TopEndpointComponent {
    /**
     * Represents an endpoint in handling top query. It only returns as many results as top arg specified.
     * @constructor TopEndpointComponent
     * @param { object } executionContext - Underlying Execution Context
     * @ignore
     */
    constructor(executionContext, topCount) {
        this.executionContext = executionContext;
        this.topCount = topCount;
    }
    /**
     * Execute a provided function on the next element in the TopEndpointComponent.
     * @memberof TopEndpointComponent
     * @instance
     * @param {callback} callback - Function to execute for each element. \
     * the function takes two parameters error, element.
     */
    nextItem() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.topCount <= 0) {
                return { result: undefined, headers: undefined };
            }
            this.topCount--;
            try {
                return this.executionContext.nextItem();
            }
            catch (err) {
                throw err;
            }
        });
    }
    /**
     * Retrieve the current element on the TopEndpointComponent.
     * @memberof TopEndpointComponent
     * @instance
     * @param {callback} callback - Function to execute for the current element. \
     * the function takes two parameters error, element.
     */
    current() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.topCount <= 0) {
                return { result: undefined, headers: undefined };
            }
            try {
                return this.executionContext.current();
            }
            catch (err) {
                throw err;
            }
        });
    }
    /**
     * Determine if there are still remaining resources to processs.
     * @memberof TopEndpointComponent
     * @instance
     * @returns {Boolean} true if there is other elements to process in the TopEndpointComponent.
     */
    hasMoreResults() {
        return this.topCount > 0 && this.executionContext.hasMoreResults();
    }
}
exports.TopEndpointComponent = TopEndpointComponent;
