// @azure/cosmos/lib/src/client/StoredProcedure/StoredProcedure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const common_1 = require("../../common");
/**
 * Operations for reading, replacing, deleting, or executing a specific, existing stored procedure by id.
 *
 * For operations to create, upsert, read all, or query Stored Procedures,
 */
class StoredProcedure {
    /**
     * Creates a new instance of {@link StoredProcedure} linked to the parent {@link Container}.
     * @param container The parent {@link Container}.
     * @param id The id of the given {@link StoredProcedure}.
     * @hidden
     */
    constructor(container, id, clientContext) {
        this.container = container;
        this.id = id;
        this.clientContext = clientContext;
    }
    /**
     * Returns a reference URL to the resource. Used for linking in Permissions.
     */
    get url() {
        return common_1.UriFactory.createStoredProcedureUri(this.container.database.id, this.container.id, this.id);
    }
    /**
     * Read the {@link StoredProcedureDefinition} for the given {@link StoredProcedure}.
     * @param options
     */
    read(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const path = common_1.Helper.getPathFromLink(this.url);
            const id = common_1.Helper.getIdFromLink(this.url);
            const response = yield this.clientContext.read(path, "sprocs", id, undefined, options);
            return { body: response.result, headers: response.headers, ref: this, storedProcedure: this, sproc: this };
        });
    }
    /**
     * Replace the given {@link StoredProcedure} with the specified {@link StoredProcedureDefinition}.
     * @param body The specified {@link StoredProcedureDefinition} to replace the existing definition.
     * @param options
     */
    replace(body, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (body.body) {
                body.body = body.body.toString();
            }
            const err = {};
            if (!common_1.Helper.isResourceValid(body, err)) {
                throw err;
            }
            const path = common_1.Helper.getPathFromLink(this.url);
            const id = common_1.Helper.getIdFromLink(this.url);
            const response = yield this.clientContext.replace(body, path, "sprocs", id, undefined, options);
            return { body: response.result, headers: response.headers, ref: this, storedProcedure: this, sproc: this };
        });
    }
    /**
     * Delete the given {@link StoredProcedure}.
     * @param options
     */
    delete(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const path = common_1.Helper.getPathFromLink(this.url);
            const id = common_1.Helper.getIdFromLink(this.url);
            const response = yield this.clientContext.delete(path, "sprocs", id, undefined, options);
            return { body: response.result, headers: response.headers, ref: this, storedProcedure: this, sproc: this };
        });
    }
    execute(params, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.clientContext.execute(this.url, params, options);
            return { body: response.result, headers: response.headers, ref: this };
        });
    }
}
exports.StoredProcedure = StoredProcedure;
