// @azure/cosmos/lib/src/test/common/TestData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @hidden */
class TestData {
    constructor(partitionKey, uniquePartitionKey) {
        this.partitionKey = partitionKey;
        this.uniquePartitionKey = uniquePartitionKey;
        this.numberOfDocuments = 800;
        this.field = "field";
        const docs = [];
        const values = [null, false, true, "abc", "cdfg", "opqrs", "ttttttt", "xyz", "oo", "ppp"];
        for (const value of values) {
            const d = {};
            d[partitionKey] = value;
            docs.push(d);
        }
        this.numberOfDocsWithSamePartitionKey = 400;
        for (let i = 0; i < this.numberOfDocsWithSamePartitionKey; ++i) {
            const d = {};
            d[partitionKey] = uniquePartitionKey;
            d["resourceId"] = i.toString();
            d[this.field] = i + 1;
            docs.push(d);
        }
        this.numberOfDocumentsWithNumbericId =
            this.numberOfDocuments - values.length - this.numberOfDocsWithSamePartitionKey;
        for (let i = 0; i < this.numberOfDocumentsWithNumbericId; ++i) {
            const d = {};
            d[partitionKey] = i + 1;
            docs.push(d);
        }
        this.sum = (this.numberOfDocumentsWithNumbericId * (this.numberOfDocumentsWithNumbericId + 1)) / 2.0;
        this.docs = docs;
    }
}
exports.TestData = TestData;
