// @azure/cosmos/lib/src/queryMetrics/runtimeExecutionTimes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const queryMetricsConstants_1 = tslib_1.__importDefault(require("./queryMetricsConstants"));
const queryMetricsUtils_1 = require("./queryMetricsUtils");
const timeSpan_1 = require("./timeSpan");
class RuntimeExecutionTimes {
    constructor(queryEngineExecutionTime, systemFunctionExecutionTime, userDefinedFunctionExecutionTime) {
        this.queryEngineExecutionTime = queryEngineExecutionTime;
        this.systemFunctionExecutionTime = systemFunctionExecutionTime;
        this.userDefinedFunctionExecutionTime = userDefinedFunctionExecutionTime;
    }
    /**
     * returns a new RuntimeExecutionTimes instance that is the addition of this and the arguments.
     */
    add(...runtimeExecutionTimesArray) {
        if (arguments == null || arguments.length === 0) {
            throw new Error("arguments was null or empty");
        }
        let queryEngineExecutionTime = this.queryEngineExecutionTime;
        let systemFunctionExecutionTime = this.systemFunctionExecutionTime;
        let userDefinedFunctionExecutionTime = this.userDefinedFunctionExecutionTime;
        for (const runtimeExecutionTimes of runtimeExecutionTimesArray) {
            if (runtimeExecutionTimes == null) {
                throw new Error("runtimeExecutionTimes has null or undefined item(s)");
            }
            queryEngineExecutionTime = queryEngineExecutionTime.add(runtimeExecutionTimes.queryEngineExecutionTime);
            systemFunctionExecutionTime = systemFunctionExecutionTime.add(runtimeExecutionTimes.systemFunctionExecutionTime);
            userDefinedFunctionExecutionTime = userDefinedFunctionExecutionTime.add(runtimeExecutionTimes.userDefinedFunctionExecutionTime);
        }
        return new RuntimeExecutionTimes(queryEngineExecutionTime, systemFunctionExecutionTime, userDefinedFunctionExecutionTime);
    }
    /**
     * Output the RuntimeExecutionTimes as a delimited string.
     */
    toDelimitedString() {
        // tslint:disable-next-line:max-line-length
        return (`${queryMetricsConstants_1.default.SystemFunctionExecuteTimeInMs}=${this.systemFunctionExecutionTime.totalMilliseconds()};` +
            // tslint:disable-next-line:max-line-length
            `${queryMetricsConstants_1.default.UserDefinedFunctionExecutionTimeInMs}=${this.userDefinedFunctionExecutionTime.totalMilliseconds()}`);
    }
    /**
     * Returns a new instance of the RuntimeExecutionTimes class that is
     *  the aggregation of an array of RuntimeExecutionTimes.
     */
    static createFromArray(runtimeExecutionTimesArray) {
        if (runtimeExecutionTimesArray == null) {
            throw new Error("runtimeExecutionTimesArray is null or undefined item(s)");
        }
        return RuntimeExecutionTimes.zero.add(...runtimeExecutionTimesArray);
    }
    /**
     * Returns a new instance of the RuntimeExecutionTimes class this is deserialized from a delimited string.
     */
    static createFromDelimitedString(delimitedString) {
        const metrics = queryMetricsUtils_1.QueryMetricsUtils.parseDelimitedString(delimitedString);
        const vmExecutionTime = queryMetricsUtils_1.QueryMetricsUtils.timeSpanFromMetrics(metrics, queryMetricsConstants_1.default.VMExecutionTimeInMs);
        const indexLookupTime = queryMetricsUtils_1.QueryMetricsUtils.timeSpanFromMetrics(metrics, queryMetricsConstants_1.default.IndexLookupTimeInMs);
        const documentLoadTime = queryMetricsUtils_1.QueryMetricsUtils.timeSpanFromMetrics(metrics, queryMetricsConstants_1.default.DocumentLoadTimeInMs);
        const documentWriteTime = queryMetricsUtils_1.QueryMetricsUtils.timeSpanFromMetrics(metrics, queryMetricsConstants_1.default.DocumentWriteTimeInMs);
        let queryEngineExecutionTime = timeSpan_1.TimeSpan.zero;
        queryEngineExecutionTime = queryEngineExecutionTime.add(vmExecutionTime);
        queryEngineExecutionTime = queryEngineExecutionTime.subtract(indexLookupTime);
        queryEngineExecutionTime = queryEngineExecutionTime.subtract(documentLoadTime);
        queryEngineExecutionTime = queryEngineExecutionTime.subtract(documentWriteTime);
        return new RuntimeExecutionTimes(queryEngineExecutionTime, queryMetricsUtils_1.QueryMetricsUtils.timeSpanFromMetrics(metrics, queryMetricsConstants_1.default.SystemFunctionExecuteTimeInMs), queryMetricsUtils_1.QueryMetricsUtils.timeSpanFromMetrics(metrics, queryMetricsConstants_1.default.UserDefinedFunctionExecutionTimeInMs));
    }
}
RuntimeExecutionTimes.zero = new RuntimeExecutionTimes(timeSpan_1.TimeSpan.zero, timeSpan_1.TimeSpan.zero, timeSpan_1.TimeSpan.zero);
exports.RuntimeExecutionTimes = RuntimeExecutionTimes;
