// @azure/cosmos/lib/src/test/integration/ruPerMin.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const assert_1 = tslib_1.__importDefault(require("assert"));
const __1 = require("../..");
const _testConfig_1 = require("../common/_testConfig");
const TestHelpers_1 = require("../common/TestHelpers");
const client = new __1.CosmosClient({ endpoint: _testConfig_1.endpoint, auth: { masterKey: _testConfig_1.masterKey } });
// TODO: these tests are all disabled
describe("RU Per Minute", function () {
    let database;
    // - removes all the databases,
    //  - creates a new database,
    beforeEach(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield TestHelpers_1.removeAllDatabases();
        database = yield TestHelpers_1.getTestDatabase("RU Per minute");
    }));
    // - removes all the databases,
    afterEach(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield TestHelpers_1.removeAllDatabases();
    }));
    xit("Create container with RU Per Minute Offer", function () {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const containerDefinition = {
                id: "sample col"
            };
            const options = {
                offerEnableRUPerMinuteThroughput: true,
                offerVersion: "V2",
                offerThroughput: 400
            };
            yield database.containers.create(containerDefinition, options);
            const { result: offers } = yield client.offers.readAll().toArray();
            assert_1.default.equal(offers.length, 1);
            const offer = offers[0];
            assert_1.default.equal(offer.offerType, "Invalid");
            assert_1.default.notEqual(offer.content, undefined);
            assert_1.default.equal(offer.content.offerIsRUPerMinuteThroughputEnabled, true);
        });
    });
    xit("Create container without RU Per Minute Offer", function () {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const containerDefinition = {
                id: "sample col"
            };
            const options = {
                offerVersion: "V2",
                offerThroughput: 400
            };
            yield database.containers.create(containerDefinition, options);
            const { result: offers } = yield client.offers.readAll().toArray();
            assert_1.default.equal(offers.length, 1);
            const offer = offers[0];
            assert_1.default.equal(offer.offerType, "Invalid");
            assert_1.default.notEqual(offer.content, undefined);
            assert_1.default.equal(offer.content.offerIsRUPerMinuteThroughputEnabled, false);
        });
    });
    xit("Create container with RU Per Minute Offer and insert Document with disableRUPerMinuteUsage options", function () {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const containerDefinition = {
                id: "sample col"
            };
            const options = {
                offerEnableRUPerMinuteThroughput: true,
                offerVersion: "V2",
                offerThroughput: 400
            };
            yield database.containers.create(containerDefinition, options);
            const container = database.container(containerDefinition.id);
            const options2 = {
                disableRUPerMinuteUsage: true
            };
            const { headers } = yield container.items.create({ id: "sample document" }, options2);
            assert_1.default(headers[__1.Constants.HttpHeaders.IsRUPerMinuteUsed] !== true);
        });
    });
});
